pkg/agent/supportbundlecollection/support_bundle_controller.go

```diff
@@ -0,0 +1,439 @@
+		watcher.Stop()
+	}()
+
+	for {
+		select {
+		case event, ok := <-watcher.ResultChan():
+			if !ok {
+				return
+			}
+			switch event.Type {
+			case watch.Bookmark:
+				klog.V(2).Info("Received Bookmark event")
+			case watch.Added:
+				c.addSupportBundleCollection(event.Object.(*cpv1b2.SupportBundleCollection))
+				klog.InfoS("Added SupportBundleCollection", "name", event.Object.(*cpv1b2.SupportBundleCollection).Name)
+			case watch.Deleted:
+				c.deleteSupportBundleCollection(event.Object.(*cpv1b2.SupportBundleCollection))
+				klog.InfoS("Deleted SupportBundleCollection", "name", event.Object.(*cpv1b2.SupportBundleCollection).Name)
+			default:
+				klog.Errorf("Unknown event: %v", event)
+				return
+			}
+			eventCount++
+		}
+
+	}
+}
+
+func (c *SupportBundleController) addSupportBundleCollection(supportbundle *cpv1b2.SupportBundleCollection) {
+	c.supportBundleCollectionsMutex.Lock()
+	defer c.supportBundleCollectionsMutex.Unlock()
+
+	c.supportBundleCollections[supportbundle.Name] = supportbundle
+	c.queue.Add(supportbundle.Name)
+}
+
+func (c *SupportBundleController) deleteSupportBundleCollection(supportbundle *cpv1b2.SupportBundleCollection) {
+	c.supportBundleCollectionsMutex.Lock()
+	defer c.supportBundleCollectionsMutex.Unlock()
+
+	delete(c.supportBundleCollections, supportbundle.Name)
+	c.queue.Add(supportbundle.Name)
+}
+
+func (c *SupportBundleController) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.InfoS("Starting", "controllerName", controllerName)
+	defer klog.InfoS("Shutting down", "controllerName", controllerName)
+
+	go wait.NonSlidingUntil(c.watchSupportBundleCollections, 5*time.Second, stopCh)
+
+	go wait.Until(c.worker, time.Second, stopCh)
+	<-stopCh
+}
+
+func (c *SupportBundleController) GetSupportBundleCollectionCondition() v1alpha1.SupportBundleCollectionConditionType {
+	c.conditionLock.RLock()
+	defer c.conditionLock.RUnlock()
+	return c.collectionCondition
+}
+
+func (c *SupportBundleController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *SupportBundleController) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	if key, ok := obj.(string); !ok {
+		klog.Errorf("Expected string in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncSupportBundleCollection(key); err == nil {
+		klog.InfoS("Successfully synced support bundle", "name", key)
+	} else {
+		// Skip retrying as the time may not meet the requirements for SupportBundle.
+		klog.ErrorS(err, "Error syncing SupportBundleCollection", "name", key)
+	}
+	return true
+}
+
+func (c *SupportBundleController) syncSupportBundleCollection(key string) error {
+	klog.InfoS("Sync support bundle collection", "name", key)
+	c.supportBundleCollectionsMutex.RLock()
+	defer c.supportBundleCollectionsMutex.RUnlock()
+	supportBundle, exists := c.supportBundleCollections[key]
+	if !exists {
+		return nil
+	}
```

Comment:
It doesn't need to lock the whole function:
```
	supportBundle := func() *cpv1b2.SupportBundleCollection {
		c.supportBundleCollectionsMutex.RLock()
		defer c.supportBundleCollectionsMutex.RUnlock()
		return c.supportBundleCollections[key]
	}()
	if supportBundle == nil {
		return nil
	}
```

Reply:
I have changed the map to an object. Since controller will always send only one supportbundle to the agent and will not send another until the current one is finished, I remove the lock for the object.

Reply:
Got it, I have corrected this part, please help review the latest version.